File: app/src/main/java/org/softeg/slartus/forpdaplus/devdb/helpers/ParseHelper.java
package org.softeg.slartus.forpdaplus.devdb.helpers;

import com.google.gson.Gson;

import org.jsoup.Jsoup;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;
import org.softeg.slartus.forpdaplus.devdb.model.CommentsModel;
import org.softeg.slartus.forpdaplus.devdb.model.DiscussionModel;
import org.softeg.slartus.forpdaplus.devdb.model.FirmwareModel;
import org.softeg.slartus.forpdaplus.devdb.model.PricesModel;
import org.softeg.slartus.forpdaplus.devdb.model.ReviewsModel;
import org.softeg.slartus.forpdaplus.devdb.model.SpecModel;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by isanechek on 23.11.15.
 */
public class ParseHelper {
    private final ParsedModel parsed = new ParsedModel();

    public ParsedModel parseHelper(String page) {
        Element main = Jsoup.parse(page).select(".device-frame").first();
        parseTitle(main);
        parseSpec(main);
        parseFirmware(main);
        parseComments(main);
        parseReviews(main);
        parseDiscussions(main);
        parsePrices(main);
        return parsed;
    }

    private void parseTitle(Element main){
        parsed.setTitle(main.select(".product-name").first().text());
    }

    private void parseSpec(Element main){
        SpecModel specModel = new SpecModel();
        Element spec = main.select("#specification").first();

        for(Element element:spec.select(".item-visual .item-gallery a")){
            specModel.getGalleryLinks().add(element.attr("href"));
            specModel.getGalleryImages().add(element.select("img").first().attr("src"));
        }
        Elements temp = spec.select(".item-main .price-box .price strong");
        if(!temp.text().isEmpty())
            specModel.setPrice(temp.first().text());
        specModel.setSpecTable(spec.select(".item-content .content .specifications-list"));
        parsed.setSpecModel(specModel);
    }
    private void parseDiscussions(Element main) {
        String link, title, time, description;
        List<DiscussionModel> cache = new ArrayList<>();
        DiscussionModel model;
        for (Element element : main.select("#discussions .article-list li")) {
            link = element.select(".title a").first().attr("href");
            title = element.select(".title").first().text();
            time = element.select(".upd").first().text();
            description = element.select(".description").first().text();
            cache.add(new DiscussionModel(description, time, link, title));
        }
        parsed.setDiscussionModels(new Gson().toJson(cache));
    }

    private void parseComments(Element main) {
        String comment, link, userName, date, ratingNum, ratingText;
        List<CommentsModel> cache = new ArrayList<>();
        ArrayList<String> dr = new ArrayList<>();
        for (Element element1 : main.select("#comments .reviews li")) {
            if (!element1.select(".text-box").text().isEmpty()) {
                /**
                 * Тут короче если текст бокс не нуль, то и все остальное не нуль.
                 */
                Element element =  element1.select(".text-box .w-toggle").first();
                if(element==null)
                    element = element1.select(".text-box").first();
                comment = element.text();
                element = element1.select("div.name a").first();
                link = element.attr("href");
                userName = element.attr("title");
                date = element1.select("div.date").first().text();
                ratingNum = element1.select("span.num").first().text();
                ratingText = element1.select("span.text").first().text();
                // for detail dialog
                Elements elements1 = element1.getElementsByClass("reviews-list");
                if (elements1 != null)
                    for (Element element2 : elements1)
                        dr.add(element2.select("div.line").text());
                cache.add(new CommentsModel(date, ratingNum, ratingText, comment, link, userName, dr));
            }
        }
        parsed.setCommentsModels(new Gson().toJson(cache));
    }


    private void parsePrices(Element main) {
        String link, title, time, description;
        List<PricesModel> cache = new ArrayList<>();
        Elements elements = main.select("#prices .article-list li");
        for(Element element:elements){
            link = element.select(".title a").first().attr("href");
            title = element.select(".title").first().text();
            time = element.select(".upd").first().text();
            description = element.select(".description").first().text();

            cache.add(new PricesModel(time, description, link, title));
        }
        parsed.setPricesModels(new Gson().toJson(cache));
    }

    private void parseFirmware(Element main) {
        String link, title, time, description;
        List<FirmwareModel> cache = new ArrayList<>();
        for(Element element:main.select("#firmware .article-list li")){
            link = element.select(".title a").first().attr("href");
            title = element.select(".title").first().text();
            time = element.select(".upd").first().text();
            description = element.select(".description").first().text();
            cache.add(new FirmwareModel(time, description, link, title));
        }
        parsed.setFirmwareModels(new Gson().toJson(cache));
    }

    private void parseReviews(Element main) {
        String url, imgLink, title, date, description;
        List<ReviewsModel> cache = new ArrayList<>();
        for(Element element:main.select("#reviews .article-list li")){
            url = element.select("a").first().attr("href");
            imgLink = element.select(".article-img img").first().attr("src");
            title = element.select(".title").first().text();
            date = element.select(".upd").first().text();
            description = element.select(".description").first().text();
            cache.add(new ReviewsModel(date, imgLink, url, description, title));
        }
        parsed.setReviewsModels(new Gson().toJson(cache));
    }
}

File: app/src/main/java/org/softeg/slartus/forpdaplus/devdb/helpers/ParseHelper.kt
package org.softeg.slartus.forpdaplus.devdb.helpers

import com.google.gson.Gson
import org.jsoup.Jsoup
import org.jsoup.nodes.Element
import org.softeg.slartus.forpdaplus.devdb.model.*

class ParseHelper {
    private val parsed = ParsedModel()
    fun parseHelper(page: String): ParsedModel {
        val main =
            Jsoup.parse(page).select(".device-frame").first() ?: error("device-frame not found")
        parseTitle(main)
        parseSpec(main)
        parseFirmware(main)
        parseComments(main)
        parseReviews(main)
        parseDiscussions(main)
        parsePrices(main)
        return parsed
    }

    private fun parseTitle(main: Element) {
        parsed.title = main.select(".product-name").first()?.text().orEmpty()
    }

    private fun parseSpec(main: Element) {
        val specModel = SpecModel()
        val spec = main.select("#specification").first() ?: return
        for (element in spec.select(".item-visual .item-gallery a")) {
            specModel.galleryLinks.add(element.attr("href"))
            specModel.galleryImages.add(element.select("img").first()!!.attr("src"))
        }
        val temp = spec.select(".item-main .price-box .price strong")
        if (temp.text().isNotEmpty())
            specModel.price = temp.first()?.text().orEmpty()
        specModel.specTable = spec.select(".item-content .content .specifications-list")
        parsed.specModel = specModel
    }

    private fun parseDiscussions(main: Element) {
        var link: String
        var title: String
        var time: String
        var description: String
        val cache: MutableList<DiscussionModel> = ArrayList()

        for (element in main.select("#discussions .article-list li")) {
            link = element.select(".title a").first()?.attr("href").orEmpty()
            title = element.select(".title").first()?.text().orEmpty()
            time = element.select(".upd").first()?.text().orEmpty()
            description = element.select(".description").first()?.text().orEmpty()
            cache.add(DiscussionModel(description, time, link, title))
        }
        parsed.discussionModels = Gson().toJson(cache)
    }

    private fun parseComments(main: Element) {
        var comment: String
        var link: String
        var userName: String
        var date: String
        var ratingNum: String
        var ratingText: String
        val cache: MutableList<CommentsModel> = ArrayList()
        val dr = ArrayList<String>()
        for (element1 in main.select("#comments .reviews li")) {
            if (!element1.select(".text-box").text().isEmpty()) {
                /**
                 * Тут короче если текст бокс не нуль, то и все остальное не нуль.
                 */
                var element = element1.select(".text-box .w-toggle").first()
                if (element == null) element = element1.select(".text-box").first()
                comment = element?.text().orEmpty()
                element = element1.select("div.name a").first()
                link = element?.attr("href").orEmpty()
                userName = element?.attr("title").orEmpty()
                date = element1.select("div.date").first()?.text().orEmpty()
                ratingNum = element1.select("span.num").first()?.text().orEmpty()
                ratingText = element1.select("span.text").first()?.text().orEmpty()
                // for detail dialog
                val elements1 = element1.getElementsByClass("reviews-list")
                for (element2 in elements1)
                    dr.add(element2.select("div.line").text())
                cache.add(CommentsModel(date, ratingNum, ratingText, comment, link, userName, dr))
            }
        }
        parsed.commentsModels = Gson().toJson(cache)
    }

    private fun parsePrices(main: Element) {
        var link: String
        var title: String
        var time: String
        var description: String
        val cache: MutableList<PricesModel> = ArrayList()
        val elements = main.select("#prices .article-list li")
        for (element in elements) {
            link = element.select(".title a").first()?.attr("href").orEmpty()
            title = element.select(".title").first()?.text().orEmpty()
            time = element.select(".upd").first()?.text().orEmpty()
            description = element.select(".description").first()?.text().orEmpty()
            cache.add(PricesModel(time, description, link, title))
        }
        parsed.pricesModels = Gson().toJson(cache)
    }

    private fun parseFirmware(main: Element) {
        var link: String
        var title: String
        var time: String
        var description: String
        val cache: MutableList<FirmwareModel> = ArrayList()
        for (element in main.select("#firmware .article-list li")) {
            link = element.select(".title a").first()?.attr("href").orEmpty()
            title = element.select(".title").first()?.text().orEmpty()
            time = element.select(".upd").first()?.text().orEmpty()
            description = element.select(".description").first()?.text().orEmpty()
            cache.add(FirmwareModel(time, description, link, title))
        }
        parsed.firmwareModels = Gson().toJson(cache)
    }

    private fun parseReviews(main: Element) {
        var url: String
        var imgLink: String
        var title: String
        var date: String
        var description: String
        val cache: MutableList<ReviewsModel> = ArrayList()
        for (element in main.select("#reviews .article-list li")) {
            url = element.select("a").first()?.attr("href").orEmpty()
            imgLink = element.select(".article-img img").first()?.attr("src").orEmpty()
            title = element.select(".title").first()?.text().orEmpty()
            date = element.select(".upd").first()?.text().orEmpty()
            description = element.select(".description").first()?.text().orEmpty()
            cache.add(ReviewsModel(date, imgLink, url, description, title))
        }
        parsed.reviewsModels = Gson().toJson(cache)
    }
}